### PR TITLE
Build a secure ESP docker image run as non-root

### PR DESCRIPTION
**Pull request by qiwzhang**
Existing docker image still run as root.  

Build a separate secure docker image with "--secure" suffix run as non-root.

Please do NOT use this image if:
1) Use ports requiring root privilege, such as port 80 inside ESP containers.
2) Use custom nginx config. The server_config path is hard coded to /etc/nginx/ folder in your nginx config.  This image moves the server_config to /home/nginx.   Please modify your nginx config before using this image.

It is also easy to make root filesystem as read-only for more security with following folders mounted externally:

/var/log/nginx, 
/var/cache/nginx
/home/nginx

Here is a GKE deployment.yaml example
```
      - name: esp
        image: gcr.io/endpoints-release/endpoints-runtime-secure:1
         args: [
           "--http_port=8081",
           "--backend=127.0.0.1:8080",
          "--service=SERVICE_NAME",
           "--rollout_strategy=managed",
         ]
       # [END esp]
         ports:
         - containerPort: 8081
        securityContext:
          readOnlyRootFilesystem: true
        volumeMounts:
          - mountPath: /var/log/nginx
            name: var-log-for-nginx
            readOnly: false
          - mountPath: /var/cache/nginx
            name: var-cache-for-nginx
            readOnly: false
          - mountPath: /home/nginx
            name: home-for-nginx
            readOnly: false

      volumes:
      - emptyDir:
          medium: Memory
        name: var-log-for-nginx          
      - emptyDir:
          medium: Memory
        name: var-cache-for-nginx
      - emptyDir:
          medium: Memory
        name: home-for-nginx        
```

### CODE REVIEW

File: docker/generic/Dockerfile [hunk 1]
```diff
@@ -26,4 +26,15 @@ EXPOSE 8090
 # since documentation still refers to using docker without additional EXPOSE step
 EXPOSE 8080
 
+# Remove the nginx user created from deb package which doesn't create the home folder.
+RUN deluser nginx
```
Comment (whitlockjc) [inline docker/generic/Dockerfile]: Docker suggests [minimizing the number of layers](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers), would it make sense to chain all of these `RUN` commands together?

Comment (qiwzhang) [reply docker/generic/Dockerfile]: Thanks.  Done

**Comment by qiwzhang** [comment]
@inklesspen Could you help to review this change?  My question is: do we want serverless image to run as non-root?   If so, doesn't its docker file need to do anything,  or the user is inherited from the base image.

**Comment by inklesspen** [comment]
Worth noting that knative mounts a volume at [`/var/log`](http://go/gh/knative/serving/blob/master/docs/runtime-contract.md#default-filesystems), meaning the `/var/log/nginx` created within the container will not exist when ESP is run in knative.

Is it possible to have the startup script ensure the existence of directories that nginx needs, perhaps before dropping privs?

**Comment by qiwzhang** [comment]
I believe deb package copied some files into /var/log/nginx.   That is why I need to wipe it out in Dockerfile and re-create it and set nginx user as its owner.

My point is: ESP is not using any files under /var/log folder,  It is ok for knative to mount it externally.

I don't know the permission of /var/log mounted for /var/log.  I hope it can be writable by user nginx. 
When I tested mounting emptyDir, (or tmpfs),  its permisstion is 777, everybody can write it.

So when I mounted  /var/log/nginx /home/nginx,  /var/cache/nginx externally, it still works

**Comment by inklesspen** [comment]
Yes, but what if you mount just `/var/log`? The `/var/log/nginx` directory exists in the Docker image, because it was created in the Dockerfile, but since `/var/log` is mounted, `/var/log/nginx` is not available.

**Comment by inklesspen** [comment]
Perhaps in start_esp we can call `ensure('/var/log/nginx')`?

**Comment by inklesspen** [comment]
Whatever changes we make here will probably work for serverless. I will try it, just to be sure, if we have a jenkins build I can use.

**Comment by inklesspen** [comment]
Verified this works for serverless.

**Comment by qiwzhang** [comment]
@inklesspen could you review it again?  I change serverless docker image to be based on -secure image.

**Comment by inklesspen** [comment]
#564 fixes the directory issue for knative

**Pull request merged**